File: core/src/io/github/jerukan/game/gameunits/unitdata/BasicUnit.java
package io.github.jerukan.game.gameunits.unitdata;

import com.badlogic.gdx.graphics.Texture;
import io.github.jerukan.game.GameState;
import io.github.jerukan.game.Player;
import io.github.jerukan.game.gameunits.Unit;
import io.github.jerukan.game.gameunits.unitdata.unitactions.*;

/** All the units that don't have anything very special to them */

public class BasicUnit extends BaseUnit {

    public BasicUnit(String name, int health, int attack, int speed, int range, int cost, BaseUnit requiredUnit, Type type, String description, Texture texture, boolean oddAnim) {
        this.name = name;
        baseHealth = health;
        baseAttack = attack;
        baseSpeed = speed;
        baseRange = range;

        baseCost = cost;

        this.requiredUnit = requiredUnit;

        this.type = type;

        this.description = description;

        if(this.type == Type.SOLDIER) {
            actions = new UnitAction[]{new MoveAction(this), new AttackAction(this), new DismissAction(this)};
            baseUpkeep = 1;
        }
        else if(this.type == Type.BUILDING) {
            actions = new UnitAction[]{new DismissAction(this)};
            baseUpkeep = 0;
        }
        else if(this.type == Type.FLYING) {
            actions = new UnitAction[]{new MoveFlyAction(this), new AttackAction(this), new DismissAction(this)};
            baseUpkeep = 1;
        }

        setTexture(texture);
        oddAnimation = oddAnim;
    }

    @Override
    public boolean canBuild(Player owner) {
        return requiredUnit == null || GameState.instance.unitState.playerHasUnit(owner, requiredUnit.id);
    }

    @Override
    public void onCreation(Unit self) {

    }

    @Override
    public void onDeath(Unit self) {

    }

    @Override
    public void onTurnStart(Unit self) {

    }

    @Override
    public void onTurnEnd(Unit self) {

    }
}

File: core/src/io/github/jerukan/game/gameunits/unitdata/BasicUnit.kt
package io.github.jerukan.game.gameunits.unitdata

import com.badlogic.gdx.graphics.Texture
import io.github.jerukan.game.GameState
import io.github.jerukan.game.Player
import io.github.jerukan.game.gameunits.Unit
import io.github.jerukan.game.gameunits.UnitRegistry
import io.github.jerukan.game.gameunits.unitdata.unitactions.*

/** All the units that don't have anything very special to them  */

class BasicUnit(name: String, health: Int, attack: Int, speed: Int, range: Int, cost: Int, requiredUnit: BaseUnit?, type: BaseUnit.Type, description: String, texture: Texture) : BaseUnit() {

    init {
        this.name = name
        baseHealth = health
        baseAttack = attack
        baseSpeed = speed
        baseRange = range

        baseCost = cost

        this.requiredUnit = requiredUnit

        this.type = type

        this.description = description

        if (this.type === BaseUnit.Type.SOLDIER) {
            actions = arrayOf(MoveAction(this), AttackAction(this), DismissAction(this))
            baseUpkeep = 1
        } else if (this.type === BaseUnit.Type.BUILDING) {
            actions = arrayOf(DismissAction(this))
            baseUpkeep = 0
        } else if (this.type === BaseUnit.Type.FLYING) {
            actions = arrayOf(MoveFlyAction(this), AttackAction(this), DismissAction(this))
            baseUpkeep = 1
        }

        this.texture = texture
    }

    override fun canBuild(owner: Player): Boolean {
        return requiredUnit == null || GameState.instance.unitManager.playerHasUnit(owner, requiredUnit!!.id)
    }

    override fun onCreation(self: Unit) {

    }

    override fun onDeath(self: Unit) {

    }

    override fun onTurnStart(self: Unit) {

    }

    override fun onTurnEnd(self: Unit) {

    }
}
